### PR TITLE
Send SetBreakpointRequest directly to bloop for workspace sources

### PR DESCRIPTION
**Pull request by marek1840**

### CODE REVIEW

File: metals/src/main/scala/scala/meta/internal/metals/debug/SourcePathProvider.scala [hunk 1]
```diff
@@ -12,6 +12,10 @@ final class SourcePathProvider(
     buildTargets: BuildTargets,
     targets: List[BuildTargetIdentifier]
 ) {
+  def isWorkspacePath(path: AbsolutePath): Boolean = {
```
Comment (olafurpg) [inline metals/src/main/scala/scala/meta/internal/metals/debug/SourcePathProvider.scala]: Should we use this instead?
https://github.com/scalameta/metals/blob/59ca771c9d1c1f3ca3d34df0be84215bc86e4a40/metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala#L257

File: metals/src/main/scala/scala/meta/internal/metals/debug/DebugProxy.scala [hunk 2]
```diff
@@ -60,9 +61,27 @@ private[debug] final class DebugProxy(
...
+      val path = adapters.adaptPath(args.getSource.getPath).toAbsolutePath
+      if (sourcePathProvider.isWorkspacePath(path)) {
+        server
+          .sendPartitioned(List(request))
+          .map(_.map(DebugProtocol.parseResponse[SetBreakpointsResponse]))
+          .map(_.head.get)
+          .map { response =>
+            val breakpoints = response.getBreakpoints
+            if (breakpoints != null) {
+              breakpoints.foreach(_.setSource(args.getSource))
```
Comment (olafurpg) [inline metals/src/main/scala/scala/meta/internal/metals/debug/DebugProxy.scala]: What are we doing here?

Comment (marek1840) [reply metals/src/main/scala/scala/meta/internal/metals/debug/DebugProxy.scala]: java-debug is not setting the sources for breakpoints, so we have to do it manually :(

**Comment by marek1840** [comment]
It seems that the [bloop breakpoint support](https://github.com/scalacenter/bloop/pull/1087) doesn't sit well with our suite of tests - it seems the jvm is not stopping at the breakpoints we specify. @jvican , could you help me understand why?

File: metals/src/main/scala/scala/meta/internal/metals/debug/DebugProxy.scala (hunk 2 above)
Comment (jvican) [reply metals/src/main/scala/scala/meta/internal/metals/debug/DebugProxy.scala]: Hmm, why is `java-debug` setting the sources for breakpoints in the former case and not now?

Comment (marek1840) [reply metals/src/main/scala/scala/meta/internal/metals/debug/DebugProxy.scala]: By former case you mean the library sources? I have to set them there as well: https://github.com/scalameta/metals/blob/master/metals/src/main/scala/scala/meta/internal/metals/debug/SetBreakpointsRequestHandler.scala#L50

**Comment by marek1840** [comment]
> How are you partitioning the requests? 

I am not - I send a "partitioning" of a single request (https://github.com/scalameta/metals/pull/1208/files#diff-734f6ad19137cce99449cbe2221fb139R68). I do so, because I still need to set the sources and no other method in the current implementation returns a `Future[Response]`. The `server.send` method is not returning a response.

Of course, once we get this working, the implementation will have to be tidied up, but for now it is sufficient workaround of a known limitation.

**Comment by jvican** [comment]
I'm not sure I fully understand your answer but are you then not following this contract " The contract java-debug has to set breakpoints correctly is that it needs to receive all breakpoints from one file together. "? The test case I linked should give you a hint of why it doesn't work, as that tests multi-breakpoints in both Scala and Java sources.

**Comment by marek1840** [comment]
> are you then not following this contract (...)

I am, I am just passing the request to Bloop as I get it. Only one request goes to bloop (for a single source, that is).

**Comment by marek1840** [comment]
As for the offline question:

> Where is the logic to read LineNumberTable in Metals?

Which, I believe was caused by me being unclear in this comment: https://github.com/scalameta/metals/pull/1163#discussion_r359731298

We are not checking this table ourselves. What we do is to resolve the fully qualified name using the source file [here](https://github.com/scalameta/metals/blob/master/metals/src/main/scala/scala/meta/internal/metals/debug/SetBreakpointsRequestHandler.scala#L23-L33) and [then use it](https://github.com/scalameta/metals/blob/master/metals/src/main/scala/scala/meta/internal/metals/debug/SetBreakpointsRequestHandler.scala#L64-L75) as a `dap-fqcn` uri.

The silent assumption here is that the compiler populates the `LineNumberTable` correctly -  I should have probably phrased that more clearly.

**Comment by jvican** [comment]
> The silent assumption here is that the compiler populates the LineNumberTable correctly - I should have probably phrased that more clearly.
  
FTR, I don't think neither the Scala nor Java compiler is using the `LineNumberTable` class file annotation.

**Comment by marek1840** [comment]
> I don't think neither the Scala nor Java compiler is using the LineNumberTable class file annotation.

Didn't know that, thanks. 
Anyway, to be more precise: we assume the jvm knows how to set the breakpoint for a specific file at line L (be it with the help of a LineNumberTable or via any other mechanism). 
My implementation assumes that lines will be resolved correctly and therefore it focuses only on resolving the fully qualified class name using the sources we have.

**Comment by jvican** [comment]
> therefore it focuses only on resolving the fully qualified class name using the sources we have.

Yes, that's my understanding. See my reply in the other PR for why we should avoid doing that for workspace sources.

**Comment by tgodzik** [comment]
Closing after discussing @marek1840 - this approach is currently not working.

**Comment by jvican** [comment]
This will be fixed in 1.4.0-RC2. I’ll comment here when the release is out.

**Pull request closed without merging**